### PR TITLE
Canonical: Preview

### PR DESCRIPTION
**Pull request by tom-wa**
# Purpose

Preview: Canonicalization plugin - matches keyvalue against either a list of values, posix extended regex or fnmatch and on a match brings the value to a canonical form

Some examples are in README.md

Pending decisions:
- move stringToArray helper function to a library ? 
- compilation variants for regex, lists and fnmatch ?
- per mountpoint configurations ? 

TODO:
- [ ] error handling
- [ ] cleanup
- [ ] Documentation

# Checklist

- [x] commit messages are fine (with references to issues)
- [x] I ran all tests and everything went fine
- [x] I added unit tests
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions
- [ ] meta data is updated (e.g. README.md of plugins)

@markus2330 please review my pull request

### CODE REVIEW

File: src/plugins/canonical/canonical.c [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+		{
+			++elems;
+		}
+		++ptr;
+	}
+	++elems; // terminating element
+	array = elektraCalloc (elems * sizeof (char *));
+	char * localString = NULL;
+	if (string[0] == '[' || string[0] == '(' || string[0] == '{')
+		localString = elektraStrDup (string + 1);
```
Comment (markus2330) [inline src/plugins/canonical/canonical.c]: minor: you could avoid the else branch by calculating an offset (which is either 0 or 1)

Please prefer to use `{}`.

File: src/plugins/canonical/canonical.c [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+	if (lastChar == ']' || lastChar == ')' || lastChar == '}') localString[strlen (localString) - 1] = '\0';
+
+	int current = 0;
+	ptr = strtok (localString, delim);
+	while (ptr)
+	{
+		if (*ptr == '\0')
+		{
+			ptr = strtok (NULL, delim);
+			continue;
```
Comment (markus2330) [inline src/plugins/canonical/canonical.c]: minor: you could easily avoid continue by putting the rest in the else branch.

File: src/plugins/canonical/canonical.c [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+		}
+		char * tmp = elektraStrDup (ptr);
+		char * start = elektraLskip (tmp);
+		(void)elektraRstrip (start, NULL);
+		if (start[0] == '\'') ++start;
+		if (start[strlen (start) - 1] == '\'') start[strlen (start) - 1] = '\0';
```
Comment (markus2330) [inline src/plugins/canonical/canonical.c]: Please use `{}`, its easy to get confused here where `[]()` start/stop.

**Comment by tom-wa** [comment]
>What about having a preconfigured way to get canonicalization like done in the plugin bool? 

i'm currently trying to figure out which approach of preconfiguring works best: plugin configuration or the way typedispatcher currently does it (mounting a configuration and letting the plugin read it)
the latter one seems easier for everything thats more than 3 configuration keys to set.
should we give plugins a way to automatically fetch configurations below some specified mountpoint on kdbOpen ?

**Comment by markus2330** [comment]
tom-wa wrote:
> i'm currently trying to figure out which approach of preconfiguring works
> best: plugin configuration or the way typedispatcher currently does it
> (mounting a configuration and letting the plugin read it) the latter one
> seems easier for everything thats more than 3 configuration keys to set.
> should we give plugins a way to automatically fetch configurations below
> some specified mountpoint on kdbOpen ?

It would be a great idea to allow plugin configs to be mounted. But it should 
not be done using kdbOpen within the plugins---we would need framework 
extensions to do this properly. And I am afraid that is out of scope for 
lcdproc or even 1.0.

For now the best solutions would be:

1. having commonly used options hard coded (such as the suggested bool 
   canonicalization, which is likely the one which is used at most by far (for 
   the tristate bool)). Your current canonical plugin config would allow such
   hard-coded plugin settings well (as option next to the regex, fnmatch, 
   list..)
2. having a type/spec-dispatcher-like plugin which copies metadata from one 
   place to all other places where needed

If the type dispatcher is already capable to copy the canonical metadata to 
all places which use the same type it would be a nice solution. Otherwise I 
would refain from bloating the type plugin even more, the review is quite hard 
because it is already a monster.

**Comment by markus2330** [comment]
I'll close it. Maybe we can continue this after Elektra 1.0.

**Pull request closed without merging**